tests/worker/test_workflow.py

```diff
@@ -3236,3 +3235,150 @@ async def test_workflow_annotated_with_self(client: Client):
+    @workflow.run
+    async def run(self) -> None:
+        await workflow.execute_activity(
+            custom_metrics_activity, schedule_to_close_timeout=timedelta(seconds=30)
+        )
+
+        histogram = workflow.metric_meter().create_histogram(
+            "my-workflow-histogram", "my-workflow-description", "my-workflow-unit"
+        )
+        histogram.record(56)
+        histogram.with_additional_attributes({"my-workflow-extra-attr": 1234}).record(
+            78
+        )
+
+
+async def test_workflow_custom_metrics(client: Client):
+    # Run worker with default runtime which is noop meter just to confirm it
+    # doesn't fail
+    async with new_worker(
+        client, CustomMetricsWorkflow, activities=[custom_metrics_activity]
+    ) as worker:
+        await client.execute_workflow(
+            CustomMetricsWorkflow.run,
+            id=f"wf-{uuid.uuid4()}",
+            task_queue=worker.task_queue,
+        )
+
+    # Create new runtime with Prom server
+    prom_addr = f"127.0.0.1:{find_free_port()}"
+    runtime = Runtime(
+        telemetry=TelemetryConfig(
+            metrics=PrometheusConfig(bind_address=prom_addr), metric_prefix="foo_"
+        )
+    )
+
+    # Confirm meter fails with bad attribute type
+    with pytest.raises(TypeError) as err:
+        runtime.metric_meter.with_additional_attributes({"some_attr": None})  # type: ignore
+    assert str(err.value).startswith("Invalid value type for key")
+
+    # New client with the runtime
+    client = await Client.connect(
+        client.service_client.config.target_host,
+        namespace=client.namespace,
+        runtime=runtime,
+    )
+
+    async with new_worker(
+        client, CustomMetricsWorkflow, activities=[custom_metrics_activity]
+    ) as worker:
+        # Record a gauge at runtime level
+        gauge = runtime.metric_meter.with_additional_attributes(
+            {"my-runtime-extra-attr1": "val1", "my-runtime-extra-attr2": True}
+        ).create_gauge("my-runtime-gauge", "my-runtime-description")
+        gauge.set(90)
+
+        # Run workflow
+        await client.execute_workflow(
+            CustomMetricsWorkflow.run,
+            id=f"wf-{uuid.uuid4()}",
+            task_queue=worker.task_queue,
+        )
+
+        # Get Prom dump
+        with urlopen(url=f"http://{prom_addr}/metrics") as f:
+            prom_str: str = f.read().decode("utf-8")
+            prom_lines = prom_str.splitlines()
+
+        # Intentionally naive metric checker
+        def matches_metric_line(
+            line: str, name: str, at_least_labels: Mapping[str, str], value: int
+        ) -> bool:
+            # Must have metric name
+            if not line.startswith(name + "{"):
+                return False
+            # Must have labels (don't escape for this test)
+            for k, v in at_least_labels.items():
+                if not f'{k}="{v}"' in line:
+                    return False
+            return line.endswith(f" {value}")
```

Comment:
More small-minded personal style suggestions while trying to digest the PR, FFTI!

```suggestion
            return all(
                f'{k}="{v}"' in line for k, v in at_least_labels.items()
            ) and line.endswith(f" {value}")
```

Reply:
Yeah, it's style preference. I do use `any`/`all` where I do think functional approach has more value.

Reply:
Yes, here I would personally say that `all(...) and cond` is more Pythonic than explicitly implementing an early-exit algorithm in place of the `all`.

Reply:
:+1: (just test code, and this style for assert is clearer with pytest because of how it expands the expression, but no assert in here so can change)